x/checkpoint/types/msg.go

```diff
@@ -0,0 +1,220 @@
+		StartBlock:      startBlock,
+		EndBlock:        endBlock,
+		RootHash:        roothash.Bytes(),
+		AccountRootHash: accountRootHash.Bytes(),
+		BorChainID:      borChainID,
+	}
+}
+
+// Type returns message type
+func (msg MsgCheckpoint) Type() string {
+	return "checkpoint"
+}
+
+func (msg MsgCheckpoint) Route() string {
+	return RouterKey
+}
+
+// GetSigners returns address of the signer
+func (msg MsgCheckpoint) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{sdk.AccAddress([]byte(msg.Proposer))}
+}
+
+func (msg MsgCheckpoint) GetSignBytes() []byte {
+	b, err := cdc.MarshalJSON(msg)
+	if err != nil {
+		panic(err)
+	}
+	return sdk.MustSortJSON(b)
+}
+
+func (msg MsgCheckpoint) ValidateBasic() error {
+	// TODO implement validation logic
+	// if bytes.Equal(msg.RootHash.Bytes(), helper.ZeroHash.Bytes()) {
+	// 	return hmCommon.ErrInvalidMsg(hmCommon.DefaultCodespace, "Invalid rootHash %v", msg.RootHash.String())
+	// }
+
+	// if msg.Proposer.Empty() {
+	// 	return hmCommon.ErrInvalidMsg(hmCommon.DefaultCodespace, "Invalid proposer %v", msg.Proposer.String())
+	// }
+
+	if msg.StartBlock >= msg.EndBlock || msg.EndBlock == 0 {
+		return hmCommon.ErrInvalidMsg
+	}
+
+	return nil
+}
+
+// GetSideSignBytes returns side sign bytes
+func (msg MsgCheckpoint) GetSideSignBytes() []byte {
+	// keccak256(abi.encoded(proposer, startBlock, endBlock, rootHash, accountRootHash, bor chain id))
+	borChainID, _ := strconv.ParseUint(msg.BorChainID, 10, 64)
+	return appendBytes32(
+		[]byte(msg.Proposer),
+		new(big.Int).SetUint64(msg.StartBlock).Bytes(),
+		new(big.Int).SetUint64(msg.EndBlock).Bytes(),
+		msg.RootHash,
+		msg.AccountRootHash,
+		new(big.Int).SetUint64(borChainID).Bytes(),
+	)
+}
+
+//
+// Msg Checkpoint Ack
+//
+
+var _ sdk.Msg = &MsgCheckpointAck{}
+
+// NewMsgCheckpointAck Add mainchain commit transaction hash to MsgCheckpointAck
+func NewMsgCheckpointAck(
+	from sdk.AccAddress,
+	number uint64,
+	proposer sdk.AccAddress,
+	startBlock uint64,
+	endBlock uint64,
+	rootHash hmCommonTypes.HeimdallHash,
+	txHash hmCommonTypes.HeimdallHash,
+	logIndex uint64,
+) MsgCheckpointAck {
+	return MsgCheckpointAck{
+		From:       from.String(),
+		Number:     number,
+		Proposer:   proposer.String(),
+		StartBlock: startBlock,
+		EndBlock:   endBlock,
+		RootHash:   rootHash.Bytes(),
+		TxHash:     txHash.Bytes(),
+		LogIndex:   logIndex,
+	}
+}
+
+func (msg MsgCheckpointAck) Type() string {
+	return "checkpoint-ack"
+}
+
+func (msg MsgCheckpointAck) Route() string {
+	return RouterKey
+}
+
+// GetSigners returns signers
+func (msg MsgCheckpointAck) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{sdk.AccAddress([]byte(msg.From))}
+
+}
+
+// GetSignBytes returns sign bytes
+func (msg MsgCheckpointAck) GetSignBytes() []byte {
+	b, err := cdc.MarshalJSON(msg)
+	if err != nil {
+		panic(err)
+	}
+	return sdk.MustSortJSON(b)
+}
+
+// ValidateBasic validate basic
+func (msg MsgCheckpointAck) ValidateBasic() error {
```

Comment:
Check implementation of ValidateBasic() methods

Reply:
Done